README.rst
audit-python-package
====================
``audit-python-package`` checks the compliance of a Python source code
repository with current packaging best practices.  It has a variety of
checks pertaining to things like:

* Versions of fundamental dependencies
* Unit test execution
* Code coverage of tests
* Supported Python version(s)
* Documentation
* Package metadata

It's implemented as a set of Python unit tests executed via pytest.  But
unlike most unit tests that are tailored to the code they're packaged with,
these ones look at the files under the directory from which they're executed.
Additional checks can be added as new test cases; doing so is usually pretty
trivial, most of the checks so far are only 1-2 lines of code each.

Currently all tests are always run, but the intent is to start customizing
this by conditionally skipping certain test cases based on the repository's
content and/or configuration settings (perhaps in the ``tox.ini`` section
added below).

Installation
------------
The easiest way to add support for auditing a package is to add a new test
environment to its ``tox.ini``::

    [testenv:audit]
    commands =
        pip --disable-pip-version-check install --upgrade --quiet audit-python-package
        py.test --pyargs audit_python_package
        python setup.py check --restructuredtext --strict --metadata

This gets the latest version of the packaging checks and then runs them.  Just
run ``tox -e audit`` from the package's root directory and start fixing the
issues that cause tests to fail (and submitting patches for any checks that do
not seem to reflect good practices).

Skipping Individual Checks
--------------------------
If some checks don't make sense for your particular package (but do for enough
others that they're still worth enabling by default), you can specifically
disable them via pytest's ``-k`` option::

    py.test --pyargs audit_python_package -k "not test_documentation_exists and not test_sbo_sphinx_version"

Tracking Dependency Updates
---------------------------
Detailed reports on the status of pinned dependency versions compared to their
latest releases can be obtained via `requires.io <https://requires.io/>`_.
These reports include Python 3 compatibility status, changelogs between the
currently used and latest available versions, and any security issues reported
for the versions currently in use.  The site can be configured to auto-scan
GitHub and Bitbucket repositories that it can access, but we'll focus here on
their API which also supports private reports on private repositories that the
requires.io site isn't allowed to access.

First, obtain an API token for your requires.io account.  Store it in a
REQUIRES_TOKEN environment variable, for example in ``.bash_profile``::

    export REQUIRES_TOKEN=1234567890abcdef

This allows using the requires.io API from scripts in any number of repositories
without needing to store the key in multiple insecure places.

Next, run the ``upload_requirements`` script from the ``audit`` tox environment
created above (be sure to run it from the project's root directory)::

    $ . .tox/audit/bin/activate
    $ upload_requirements
    $ deactivate

This creates (if it doesn't already exist) a private repository on requires.io
and uploads any ``requirements/*.txt`` files (other than
``uninstall.txt``) for the current git branch.  If you do this on a temporary
branch (for a pull request, etc.), you'll probably want to delete the branch's
entry from the requires.io UI once the real branch has been deleted.  To do
this, run the following command with the ``audit`` virtualenv active::

    requires.io delete-branch -r <repository_name> -n <branch_name>

Patches to also support similar dependency tracking services such as
`VersionEye <https://www.versioneye.com/>`_ and
`Gemnasium <https://gemnasium.com/>`_ are welcome.


docs/CHANGELOG.rst
audit-python-package Changelog
==============================

1.7.6 (2016-03-21)
------------------
Updated recommended versions of CommonMark, pip, pytest, pytz, setuptools,
traitlets, and virtualenv.

1.7.5 (2016-03-02)
------------------
Updated recommended versions of decorator, ipdb, ipython, pexpect, pip,
Pygments, pytest, pytest-cov, setuptools, Sphinx, and virtualenv.

1.7.4 (2016-01-29)
------------------
Updated recommended versions of setuptools and virtualenv

1.7.3 (2016-01-27)
------------------
* Reversed the check for explicit inclusion of exact dependency versions for
  doc builds in setup.py; this turned out to be prone to version conflicts
  between different installed packages which all do this.  Instead, you should
  specify a requirements file for the virtualenv in your Read the Docs
  configuration for the project.
* Added checks for specifying exact versions of CommonMark and recommonmark,
  which are indirect dependencies of sbo-sphinx.  As of this writing, the
  latest recommonmark is not compatible with the latest CommonMark release;
  see https://github.com/rtfd/recommonmark/issues/24 .

1.7.2 (2016-01-26)
------------------
* Accommodate the `readme` -> `readme_renderer` package rename
* Updated recommended versions of alabaster, Babel, decorator, ipython,
  pickleshare, pip, Pygments, pytest, pytest-catchlog, setuptools,
  snowballstemmer, Sphinx, tox, traitlets, and virtualenv.  Note that the
  pip upgrade may require some parameter changes if installing packages from
  private package indexes.

1.7.1 (2015-12-07)
------------------
* Updated recommended versions of coverage, ipython, py, pytest, setuptools,
  and Sphinx

1.7.0 (2015-11-18)
------------------
* Added "readme" as a dependency (for PyPI description validation)
* Removed dependency on "future", since "six" is smaller and was needed for
  "readme" anyway
* Updated the recommended version of tox again for a bugfix release
* Fixed a bug in the Python 3.5 test environment check

1.6.9 (2015-11-11)
------------------
* Updated recommended versions of coverage, pytest-catchlog, pytz, setuptools,
  and tox
* Check for a Python 3.5 test environment in tox.ini instead of 3.4

1.6.8 (2015-10-22)
------------------
Updated recommended versions of coverage and requests.

1.6.7 (2015-10-09)
------------------
* Updated recommended versions of path.py, pytest, pytest-cov, setuptools, and
  six.
* Expanded docstrings to better explain the reasoning behind the packaging
  tests.

1.6.6 (2015-10-01)
------------------
* Updated recommended versions of Babel, bleach, coverage, decorator, html5lib,
  path.py, pluggy, pytest, pytz, setuptools, sbo-sphinx, and sphinx_rtd_theme
* Added .cache directory to .gitignore recommendations to accommodate
  coverage 4.0

1.6.5 (2015-08-27)
------------------
* Updated recommended versions of ipython, pip, pytest-cov readme, setuptools,
  and virtualenv.
* Added tests for new dependencies of ipython 4.0.0: appnope, decorator,
  ipython-genutils, path.py, pexpect, pickleshare, simplegeneric, and traitlets

1.6.4 (2015-08-03)
------------------
Updated recommended versions of Babel, Jinja2, and pytest-cov.

1.6.3 (2015-07-16)
------------------
Updated checks to encourage usage of a ``requirements/tox.txt`` file for tox
and its dependencies, instead of keeping them in ``requirements/tests.txt``.
This makes it easier to keep dependencies current in Jenkins or Travis jobs
without constantly going back to edit the job configuration.

1.6.2 (2015-07-15)
------------------
* Check to makes sure that if setup.py's install_requires is derived from a
  requirements file, it respects any environment markers the file may contain
* Updated recommended versions of html5lib and ipython
* Upgraded requires.io package to 0.2.5
* Improved organization of requirements file checks

1.6.1 (2015-07-02)
------------------
* Updated recommended versions of pip, py, setuptools, sbo-sphinx, and
  virtualenv
* Added tox environment for flake8 checks, fixed reported issues

1.6.0 (2015-06-25)
------------------
* Added check for the presence of include_package_data=True and the absence of
  package_data in setup.py
* Updated checks for tox dependencies to reflect the fact that it now depends
  on pluggy
* Added checks for ipdb and its dependencies being in requirements/tests.txt
* Don't choke on environment markers when parsing requirements files
* Recurse through included requirements files when looking for a package
  specification in a requirements file
* Added checks to verify that cpython2.txt, pypy.txt, and cpython3.text
  requirements files are not present (environment markers in requirements
  files are a better way to handle this)
* Added check for an invalid 'Private :: Do Not Upload' classifier in setup.py
  to prevent accidental release of private packages to the public PyPI server
* Updated recommended package versions to reflect recent releases

1.5.0 (2015-06-05)
------------------
* Added tests for requirements/clean_up_requirements.py and usage of it in
  git-hooks/post-merge and tox.ini
* Added tests for requirements/uninstall.txt
* Upgraded setuptools, pip, pytest-catchlog, tox, and virtualenv recommended
  versions

1.4.2 (2015-05-29)
------------------
Added requires.io requirement (0.2.4) to setup.py.

1.4.1 (2015-05-28)
------------------
* Updated several recommended versions
* Removed setup.py from files uploaded to requires.io (the bug in their library
  that required it has been fixed)

1.4.0 (2015-05-18)
------------------
* Added upload_requirements script to easily track dependencies in requires.io

1.3.0 (2015-05-06)
------------------
* Added check for reporting of lines not covered by tests
* Switch post-merge check to recommend fetching setuptools & pip versions
  from requirements/base.txt
* Allow for pip parameters like "--trusted-host" in git-hooks/post-merge

1.2.1 (2015-04-30)
------------------

audit_python_package/tests/test_tox.py
# encoding: utf-8
# Created by Jeremy Bowman on Thu Apr 23 10:36:32 EDT 2015
# Copyright (c) 2015 Safari Books Online. All rights reserved.

from __future__ import unicode_literals

import os
import re

import pytest

from audit_python_package import parse_config_file


@pytest.fixture(scope='module')
def tox_ini():
    """Fixture containing the parsed content of tox.ini"""
    return parse_config_file('tox.ini')


@pytest.fixture(scope='module')
def docs_commands(tox_ini):
    """Fixture containing the commands entry from [testenv:docs]"""
    if not tox_ini.has_option('testenv:docs', 'commands'):
        return []
    return [line.strip() for line in tox_ini.get('testenv:docs', 'commands').split('\n')]


@pytest.fixture(scope='module')
def testenv_commands(tox_ini):
    """Fixture containing the commands entry from [testenv]"""
    if not tox_ini.has_option('testenv', 'commands'):
        return []
    return [line.strip() for line in tox_ini.get('testenv', 'commands').split('\n')]


class TestTox(object):
    """
    Checks related to tox.ini.  This is the configuration file for tox, which
    is used to isolate dependencies for various tasks and test configurations.
    It also contains some configuration for pytest, a popular test runner.
    """

    def test_tox_ini_exists(self):
        """There should be a tox.ini in the project's root directory"""
        assert os.path.exists('tox.ini')

    def test_tox_section_exists(self, tox_ini):
        """There should be a [tox] section in tox.ini"""
        assert tox_ini.has_section('tox')

    def test_no_downloadcache(self, tox_ini):
        """There should be no downloadcache line in tox.ini (it's deprecated)"""
        assert not tox_ini.has_option('tox', 'downloadcache')

    def test_python_35(self, tox_ini):
        """Python 3.5 should be among the tested configurations"""
        assert tox_ini.has_option('tox', 'envlist')
        assert '35' in tox_ini.get('tox', 'envlist')

    def test_pytest_section_exists(self, tox_ini):
        assert tox_ini.has_section('pytest')

    def test_coverage_modules_specified(self, tox_ini):
        """The pytest section should have an addopts entry which specifies the modules of interest for test coverage"""
        assert tox_ini.has_option('pytest', 'addopts')
        assert re.search(r'--cov \w+', tox_ini.get('pytest', 'addopts'))

    def test_coverage_term_missing(self, tox_ini):
        """The pytest section should have an addopts entry which specifies that the line numbers for missing coverage should be reported"""
        assert tox_ini.has_option('pytest', 'addopts')
        assert '--cov-report term-missing' in tox_ini.get('pytest', 'addopts')

    def test_norecursedirs(self, tox_ini):
        """Directories which shouldn't be searched for tests should be specified"""
        assert tox_ini.has_option('pytest', 'norecursedirs')
        dir_list = tox_ini.get('pytest', 'norecursedirs').split()
        assert '.*' in dir_list
        assert 'docs' in dir_list
        assert 'requirements' in dir_list
        assert 've' in dir_list

    def test_testenv_section_exists(self, tox_ini):
        """There should be a [testenv] section in tox.ini"""
        assert tox_ini.has_section('testenv')

    def test_deps(self, tox_ini):
        """There should be a "deps" line for dependencies under [testenv]"""
        assert tox_ini.has_option('testenv', 'deps')

    def test_testenv_cleans_up_requirements(self, testenv_commands):
        """There should be a command in testenv to run requirements/clean_up_requirements.py"""
        assert '{toxinidir}/requirements/clean_up_requirements.py' in testenv_commands

    def test_testenv_installs_core_dependencies(self, testenv_commands):
        """There should be a command in testenv to install the core dependencies from base.txt in the requirements directory"""
        assert any([re.match(r'pip .*install .*--requirement requirements/base.txt.*', command) for command in testenv_commands])

    def test_testenv_installs_testing_dependencies(self, testenv_commands):
        """There should be a command in testenv to install the testing dependencies from requirements/tests.txt"""
        assert any([re.match(r'pip .*install .*--requirement requirements/tests.txt.*', command) for command in testenv_commands])

    def test_testenv_uses_pytest(self, testenv_commands):
        """pytest should be the default test runner"""
        assert any([command.startswith('py.test') for command in testenv_commands])

    def test_pytest_uses_posargs(self, testenv_commands):
        """The py.test command should accept positional arguments for running specific tests"""
        assert any([re.search(r'py\.test [^\n]*\{posargs', command) for command in testenv_commands])

    def test_docs_section_exists(self, tox_ini):
        """There should be a [testenv:docs] section in tox.ini"""
        assert tox_ini.has_section('testenv:docs')

    def test_docs_cleans_up_requirements(self, docs_commands):
        """The docs test environment should run requirements/clean_up_requirements.py"""
        assert '{toxinidir}/requirements/clean_up_requirements.py' in docs_commands

    def test_docs_installs_core_dependencies(self, docs_commands):
        """The docs test environment should install the core dependencies from base.txt in the requirements directory"""
        assert any([re.match(r'pip .*install .*--requirement requirements/base.txt.*', command) for command in docs_commands])

    def test_docs_installs_documentation_dependencies(self, docs_commands):
        """The docs test environment should install the doc generation dependencies from requirements/documentation.txt"""
        assert any([re.match(r'pip .*install .*--requirement requirements/documentation.txt.*', command) for command in docs_commands])

    def test_docs_check_readme(self, docs_commands):
        """The docs test environment should include a validation of README.rst"""
        assert 'python setup.py check --restructuredtext --strict' in docs_commands

    def test_docs_sphinx_build(self, docs_commands):
        """The docs test environment should include a sphinx-build command"""
        assert 'sphinx-build -b {posargs:html} docs docs/_build' in docs_commands


audit_python_package/tests/test_docs.py
# encoding: utf-8
# Created by Jeremy Bowman on Mon Apr 27 13:55:23 EDT 2015
# Copyright (c) 2015 Safari Books Online. All rights reserved.

from __future__ import unicode_literals

import os

import pytest

from audit_python_package import get_file_content


@pytest.fixture(scope='module')
def conf():
    """Fixture for the text content of docs/conf.py"""
    return get_file_content(os.path.join('docs', 'conf.py'))


@pytest.fixture(scope='module')
def index():
    """Fixture for the text content of docs/index.rst"""
    return get_file_content(os.path.join('docs', 'index.rst'))


class TestDocs(object):
    """
    Tests of the "docs" folder for documentation.  This is the usual location
    of easy-to-edit text documentation files to be converted by Sphinx into
    easier-to-read-and-search documentation in HTML, EPUB, PDF, or other
    formats.  There are other ways to write documentation, but this one is by
    far the most common and well understood in the Python community.

    These tests additionally recommend use of the sbo-sphinx library, which
    simplifies the conf.py module and makes it possible to auto-generate
    Python API documentation from the source code during Read the Docs
    documentation builds (among other benefits).
    """

    def test_conf_exists(self):
        """There should be a docs/conf.py file with basic configuration info for the docs"""
        assert os.path.exists(os.path.join('docs', 'conf.py'))

    def test_sbo_sphinx_imports(self, conf):
        """The default configuration from sbo-sphinx should be imported in docs/conf.py"""
        assert 'from sbo_sphinx.conf import *' in conf

    def test_project_name_specified(self, conf):
        """The "project" variable should be set in docs/conf.py"""
        assert 'project = ' in conf

    def test_apidoc_exclude(self, conf):
        """The "apidoc_exclude" variable should be set in docs/conf.py to a list of modules to exclude from the API documentation"""
        assert 'apidoc_exclude = [' in conf

    def test_index_exists(self):
        """There should be a docs/index.rst file for the documentation main page"""
        assert os.path.exists(os.path.join('docs', 'index.rst'))

    def test_index_has_toctree(self, index):
        """docs/index.rst should include a "toctree" directive"""
        assert '.. toctree::' in index

    def test_api_docs_in_index(self, index):
        """docs/index.rst should include an entry for the Python API documentation"""
        assert '<python/modules>' in index

    def test_general_index_in_index(self, index):
        """docs/index.rst should include an entry for the general index ("genindex")"""
        assert ':ref:`genindex`' in index

    def test_module_index_in_index(self, index):
        """docs/index.rst should include an entry for the module index ("modindex")"""
        assert ':ref:`modindex`' in index

    def test_search_in_index(self, index):
        """docs/index.rst should include an entry for the search page"""
        assert ':ref:`search`' in index

    def test_readme_exists(self):
        """There should be a docs/readme.rst referencing the root directory's README.rst"""
        path = os.path.join('docs', 'readme.rst')
        assert os.path.exists(path)
        with open(path, 'rU') as f:
            assert '.. include:: ../README.rst' in f.read()

    def test_readme_in_index(self, index):
        """There should be an entry in docs/index.rst for docs/readme.rst"""
        assert 'readme' in index

    def test_changelog_exists(self):
        """There should be a docs/CHANGELOG.rst file for the change history"""
        assert os.path.exists(os.path.join('docs', 'CHANGELOG.rst'))

    def test_changelog_in_index(self, index):
        """There should be an entry in docs/index.rst for docs/CHANGELOG.rst"""
        assert 'CHANGELOG' in index


audit_python_package/tests/test_setup.py
# encoding: utf-8
# Created by Jeremy Bowman on Thu Apr 23 10:36:32 EDT 2015
# Copyright (c) 2015 Safari Books Online. All rights reserved.

from __future__ import unicode_literals

import os
import re

import pytest

from audit_python_package import get_file_content


@pytest.fixture(scope='module')
def setup():
    """Fixture containing the text content of setup.py"""
    return get_file_content('setup.py')


class TestSetup(object):
    """Checks related to setup.py"""

    def test_exists(self):
        """There should be a setup.py in the project's root directory"""
        assert os.path.exists('setup.py')

    def test_changelog_reminder(self, setup):
        """There should be a reminder in setup.py to update docs/CHANGELOG.rst when the version changes"""
        assert 'docs/CHANGELOG.rst' in setup

    def test_environment_markers(self, setup):
        """If install_requires is derived from a requirements file, it should respect environment markers"""
        if 'parse_requirements' in setup:
            assert '.match_markers()' in setup

    def test_include_package_data(self, setup):
        """include_package_data should be set to True in setup.py (to let MANIFEST.in define the data to include)"""
        assert 'include_package_data=True' in setup

    def test_long_description(self, setup):
        """The package's long_description should be set to the content of README.rst"""
        assert 'import codecs' in setup
        assert "with codecs.open('README.rst', 'r', 'utf-8') as f:" in setup
        assert 'long_description = f.read()' in setup
        assert 'long_description=long_description,' in setup

    def test_no_package_data(self, setup):
        """package_data should not be set in setup.py (use MANIFEST.in instead)"""
        assert not re.search(r'\spackage_data', setup)

    def test_prevent_pypi_upload(self, setup):
        """There should be an invalid "Private :: Do Not Upload" classifier in private packages to prevent accidental uploads to PyPI"""
        assert 'Private :: Do Not Upload' in setup

    def test_no_read_the_docs_section(self, setup):
        """There should not be explicit support for Read the Docs builds in setup.py (specify a requirements file for the virtualenv in the Read the Docs configuration instead)"""
        assert 'READTHEDOCS' not in setup
        assert 'documentation.txt' not in setup

    def test_multiple_pip_versions(self, setup):
        """setup.py should work with a good variety of pip versions"""
        if 'parse_requirements' in setup:
            assert 'PipSession()' in setup
            assert ', session=session)' in setup
        assert '[r.req' not in setup


audit_python_package/__init__.py
from __future__ import unicode_literals

import codecs
from six.moves.configparser import ConfigParser
import os

DATA_DIRECTORY_PATH = os.path.abspath(os.path.join(os.path.dirname(__file__), 'data'))


def parse_config_file(path):
    """Get the parsed content of an INI-style config file (using ConfigParser)"""
    config = ConfigParser()
    if os.path.exists(path):
        config.read(path)
    return config


def get_file_content(path):
    """Get the content of the UTF-8 text file at the specified path.
    Used for pytest fixtures."""
    if not os.path.exists(path):
        return ''
    with codecs.open(path, 'r', 'utf-8') as f:
        return f.read()


def get_file_lines(path):
    """Get a list of the lines in the UTF-8 text file at the specified path.
    Strips leading and trailing whitespace from each line.
    Used for pytest fixtures."""
    if not os.path.exists(path):
        return []
    with codecs.open(path, 'r', 'utf-8') as f:
        return [line.strip() for line in f.readlines()]


def get_requirement_lines(path):
    """Get the lines from a requirements file (or any other requirements file
    it includes) which contain a requirement specification."""
    lines = get_file_lines(path)
    result = []
    for line in lines:
        if line.startswith('#'):
            continue
        if line.startswith('-r') or line.startswith('--requirement'):
            base_dir = os.path.dirname(path)
            sub_path = os.path.join(base_dir, line.split()[-1])
            result.extend(get_requirement_lines(sub_path))
            continue
        if '==' not in line:
            continue
        line = line.partition(';')[0].strip()
        result.append(line)
    return result


def _get_versions_dict():
    """Parse the preferred versions list in data/requirements.txt into a
    dictionary for ease of use in tests"""
    path = os.path.join(DATA_DIRECTORY_PATH, 'requirements.txt')
    lines = get_requirement_lines(path)
    result = {}
    for line in lines:
        name, version = (part.strip() for part in line.split('=='))
        result[name] = version
    return result

VERSIONS = _get_versions_dict()


requirements/clean_up_requirements.py
#!/usr/bin/env python
# encoding: utf-8
# Created by Jeremy Bowman on Wed May 13 14:36:58 EDT 2015
# Copyright (c) 2015 Safari Books Online, LLC. All rights reserved.

"""
Utility to uninstall any installed packages listed in uninstall.txt.  Can't use
``pip uninstall -r`` because it aborts after the first package in the list
which isn't currently installed.  Also installs the currently-recommended
versions of setuptools and pip from requirements/base.txt (to avoid needing
to replicate this logic in git-hooks/post-merge, tox.ini, quilter, etc.)
"""

from __future__ import print_function, unicode_literals

import codecs
import os
import re

from pip.download import PipSession
from pip.exceptions import UninstallationError
from pip.index import PackageFinder
from pip.req import parse_requirements


if __name__ == '__main__':
    requirements_dir = os.path.abspath(os.path.dirname(__file__))
    requirements_path = os.path.join(requirements_dir, 'uninstall.txt')

    session = PipSession()
    finder = PackageFinder([], [], session=session)
    requirements = parse_requirements(requirements_path, finder, session=session)
    print('Uninstalling former requirements...')
    for req in requirements:
        try:
            req.uninstall(auto_confirm=True)
            print('Uninstalled {}'.format(req.name))
        except UninstallationError as e:
            if 'not installed' in str(e):
                # Good, the whole point was that we don't want it installed
                pass
            else:
                # We probably want to hear about any other uninstallation problems
                raise
    print('Done')

    # Install the currently recommended setuptools and pip versions
    with codecs.open('requirements/base.txt', 'r', 'utf-8') as f:
        requirements = f.read()
    match = re.search(r'^setuptools==([\d\.]+)$', requirements, re.MULTILINE)
    os.system('pip install setuptools=={}'.format(match.group(1)))
    match = re.search(r'^pip==([\d\.]+)$', requirements, re.MULTILINE)
    os.system('pip install pip=={}'.format(match.group(1)))


audit_python_package/command_line.py
# encoding: utf-8
# Created by Jeremy Bowman on Thu May 14 10:23:09 EDT 2015
# Copyright (c) 2015 Safari Books Online. All rights reserved.

"""Command line scripts to be installed by setup.py"""

from __future__ import print_function, unicode_literals

import os
import re
from subprocess import CalledProcessError, check_output, STDOUT
import sys

from audit_python_package import get_file_content


def upload_requirements():
    """
    Command line utility to upload requirements files to requires.io.  Can't do
    this easily via tox commands because it doesn't have good support for using
    the output of shell expressions (like one to get the current git branch) as
    command line parameters.  Assumes that a REQUIRES_TOKEN environment
    variable has been set to the requires.io API access token, and that it's
    being run from the project's root directory (the parent of the
    ``requirements`` directory).
    """
    verify_requires_token()
    repository_name = get_repository_name()
    update_repo(repository_name)
    branch_name = get_branch_name()
    paths = requirements_file_paths()
    update_branch(repository_name, branch_name, paths)


def verify_requires_token():
    """Make sure that the REQUIRES_TOKEN environment variable has been set"""
    requires_token = os.getenv('REQUIRES_TOKEN')
    if not requires_token:
        print('REQUIRES_TOKEN environment variable must be set')
        sys.exit(1)


def get_repository_name():
    """Get the name of the repository from setup.py"""
    setup = get_file_content('setup.py')
    match = re.search(r'name=[\'"]([^\'"]+)[\'"]', setup)
    if not match:
        print('Could not find repository name in setup.py')
        sys.exit(1)
    else:
        return match.group(1)


def get_branch_name():
    """Get the name of the current git branch"""
    try:
        return check_output(['git', 'rev-parse', '--abbrev-ref', 'HEAD'], universal_newlines=True).strip()
    except Exception as e:
        print('Error getting current git branch: {}'.format(e))
        print(e.output)
        sys.exit(1)


def requirements_file_paths():
    """List the paths relative to the project root of all requirements files
    that should be uploaded to requires.io.  Excludes
    ``requirements/uninstall.txt``."""
    paths = []
    for filename in os.listdir('requirements'):
        if len(filename) < 5 or filename[-4:] != '.txt' or filename == 'uninstall.txt':
            continue
        path = os.path.join('requirements', filename)
        paths.append(path)
    return paths


def update_repo(repository_name):
    """Create or update a private repository entry in requires.io"""
    try:
        output = check_output(['requires.io', 'update-repo', '--repository', repository_name, '--private'],
                              stderr=STDOUT, universal_newlines=True)
        print(output)
    except CalledProcessError as e:
        print('Unable to create or update {} on requires.io'.format(repository_name))
        print(e.output)
        sys.exit(1)


def update_branch(repository_name, branch_name, paths):
    """Upload the requirements files for the specified git branch to requires.io
    for analysis"""
    args = ['requires.io', 'update-branch', '--repository', repository_name, '--name', branch_name]
    for path in paths:
        args.append(path)
    try:
        output = check_output(args, stderr=STDOUT, universal_newlines=True)
        print(output)
    except CalledProcessError as e:
        print('Unable to update branch {} on requires.io'.format(branch_name))
        print(e.output)
        sys.exit(1)
